pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -2241,6 +2254,99 @@ func (r *ReconcileClusterDeployment) addOwnershipToSecret(cd *hivev1.ClusterDepl
+		return nil
+	}
+	secretName := cd.Spec.Provisioning.InstallConfigSecretRef.Name
+	log = log.WithField("installconfig-secret", secretName)
+	icSecret := &corev1.Secret{}
+	if err := r.Get(context.Background(),
+		types.NamespacedName{
+			Namespace: cd.Namespace,
+			Name:      secretName,
+		},
+		icSecret); err != nil {
+		log.WithError(err).Error("Failed to load install-config secret. We'll try again later.")
+		return nil
+	}
+	ic, err := ValidateInstallConfig(cd, icSecret)
+	if err != nil {
+		log.WithError(err).Warn("Couldn't validate install-config")
+		// This doesn't necessarily mean we didn't unmarshall it, though. Carry on...
+	}
+	if ic == nil {
+		// Today this only happens if unmarshalling fails. Otherwise we've got *something* for an install-config.
+		log.Warn("Did not get an install-config from the secret, but also got no error. This should not happen.")
+		return nil
+	}
+
+	return ic
+}
+
+// discoverAWSHostedZoneRole is intended to retrofit AWS clusters which were a) adopted, or b) created before
+// support for shared VPC was added. It attempts to discover the hosted zone role for the cluster by parsing
+// the install-config.
+// No op if platform is not AWS.
+// No op if the field is already set (we don't validate that it is correct).
+// This is best effort. If things go wrong, we log but do not fail out. THIS WILL BLOCK DEPROVISIONING. If
+// this happens, the field must be set manually.
+// The return indicates whether the field was modified. If `true`, the caller is responsible for pushing
+// an Update() back to the server.
+func (r *ReconcileClusterDeployment) discoverAWSHostedZoneRole(cd *hivev1.ClusterDeployment, log log.FieldLogger) bool {
+	log = log.WithField("subtask", "discoverAWSHostedZoneRole")
+
+	// Do we need to do this at all?
+	if getClusterPlatform(cd) != constants.PlatformAWS {
+		return false
+	}
+	if !cd.Spec.Installed {
+		// Provisioner should set the field when installation finishes; no-op for now.
+		return false
+	}
+	if controllerutils.AWSHostedZoneRole(cd) != nil {
+		// Already set, nothing to do
+		return false
+	}
+
+	// Initialize structs as necessary
+	if cd.Spec.ClusterMetadata == nil {
+		cd.Spec.ClusterMetadata = &hivev1.ClusterMetadata{}
+	}
+	if cd.Spec.ClusterMetadata.Platform == nil {
+		cd.Spec.ClusterMetadata.Platform = &hivev1.ClusterPlatformMetadata{}
+	}
+	if cd.Spec.ClusterMetadata.Platform.AWS == nil {
+		cd.Spec.ClusterMetadata.Platform.AWS = &aws.Metadata{}
+	}
+
+	// Parse the install-config
+	ic := r.getInstallConfig(cd, log)
+	if ic == nil {
+		// getInstallConfig logged
+		return false
+	}
+
+	if ic.Platform.AWS == nil {
+		// This log might be redundant with one from ValidateInstallConfig
+		log.Warn("No AWS platform section in install-config")
+		return false
+	}
+
+	hzr := ic.Platform.AWS.HostedZoneRole
+	// It's okay if this is the empty string.
+	log.WithField("hostedZoneRole", hzr).Info("Found AWS HostedZoneRole in install-config")
```

Comment:
If this cluster is non shared VPC wont we log the empty string saying the hosted zone role "" was found?

Reply:
Yup.

Reply:
I should elaborate: That's useful information, particularly in this code path where we're retrofitting existing clusters.

Reply:
I was thinking that might be confusing to someone debugging issues. Not sure if we want to have two messages one where its found (not an empty string) and one where the role is not set, though might be confusing with the `nil` case. Up to you.